README.txt
============
Introduction
============

.. image:: https://travis-ci.org/collective/collective.contentrules.mail.svg?branch=master
       :target: https://travis-ci.org/collective/collective.contentrules.mail

This package provides an extended content rule "mail" action with many more
variable substitutions available, including the name and email address of
the content owner, making it possible to email the owner of a piece of
content.

It is also possible to provide new "replacers" by writing an interface with
an associated adapter and registering a new `IMailModel` utility. The user
is able to pick the replacer to use on the action edit form.

Compatibility
=============

Plone 4.0 - 4.3


CHANGES.rst
Changelog
=========

0.8 (unreleased)
----------------

- Nothing changed yet.


0.7 (2016-03-17)
----------------

- Ported tests to plone.app.testing
  [tomgross]

- Plone 4.3 compatibiliy
  [tomgross]

- Moved to https://github.com/collective/collective.contentrules.mail
  [maurits]

0.6 (2012-02-23)
----------------

* Remove an extra From header when calling mailhost.secureSend()
  [dimboo]
  
0.5 - 2011-05-20
----------------

* Added an exception in IVocabularyFactory import for Plone >= 4.1.
  [lepri]

0.4 - 2009-09-06
----------------

* Avoid a site error if the mail cannot be sent, e.g. because the mail
  server is down.
  [optilude]

* Avoid throwing an exception on encoded strings in the replacer.
  [optilude]

* Caught workflow exception when trying to get state on object without any
  workflow.
  [tiazma]

* Added cc and bcc fields to IMailAction enabling the use of these fields in 
  mail messages
  [erico_andrei]

* Added new variables to IMailReplacer so it is possible to send emails to 
  users with other roles than owner on the object
  [erico_andrei]

* Made sure Archetypes content implements IDublinCore (It's not the default 
  behavior until AT 1.5.10)
  [erico_andrei]
  
* Correct action mail fail on try to replace bcc or cc when it returns None [lucmult]

0.3 - 2009/03/24
----------------

* Added a note in the UI about how to use variables.
  [optilude]
  
* The 'source' field is not required, but would throw an error if not filled
  in. This is now fixed.
  [optilude]

0.2 - 2009/03/24
----------------

* Made the executor fail gracefully if the content being acted upon is not
  adaptable to the replacer interface.
  [optilude]

* Made the default replacer adapt any Dublin Core-capable content, avoiding
  an unnecessary dependency on Archetypes.
  [optilude]

* Made the default IMailModel implementation more re-usable.
  [optildue]

* Updated documentation and interface specifications.
  [optilude]

* Fixes actions/configure.zcml in order to play nice with GenericSetup


0.1 - Unreleased
----------------

* Initial release



collective/contentrules/mail/tests/base.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""" defines a test class and its Plone Site layer for plone tests
"""
from plone.app.testing.bbb import PloneTestCaseFixture, PloneTestCase
from plone.app import testing
from plone.app.testing import setRoles, TEST_USER_ID
from plone.testing import z2

# fake mailhost
from Products.MailHost import MailHost


class TestMailHost(MailHost.MailHost):

    def _send(self, mfrom, mto, messageText):
        """Fake sender"""
        print messageText


class CCMFixture(PloneTestCaseFixture):

    def setUpZope(self, app, configurationContext):
        super(PloneTestCaseFixture, self).setUpZope(app, configurationContext)
        import collective.contentrules.mail
        self.loadZCML(package=collective.contentrules.mail)
        z2.installProduct(app, 'collective.contentrules.mail')

    def setUpPloneSite(self, portal):
        super(PloneTestCaseFixture, self).setUpPloneSite(portal)

        # prepare structure
        if 'news' not in portal:
            setRoles(portal, TEST_USER_ID, ['Manager'])
            portal.invokeFactory('Folder', 'news', title='News Folder')
            setRoles(portal, TEST_USER_ID, ['Member'])

        # install product
        testing.applyProfile(portal, 'collective.contentrules.mail:default')

        # install testing profile
        from Products.GenericSetup import EXTENSION, profile_registry
        profile_registry.registerProfile('testing',
            "collective.contentrules.mail testing",
            'Used for testing only',
            'profiles/testing',
            'collective.contentrules.mail',
            EXTENSION)
        testing.applyProfile(portal, 'collective.contentrules.mail:testing')
        # patch mail host
        #portal._old = MailHost.MailHost
        #MailHost.MailHost = TestMailHost

        portal.manage_changeProperties(email_from_name='Site Administrator')

    def tearDownZope(self, app):
        super(PloneTestCaseFixture, self).tearDownZope(app)
        z2.uninstallProduct(app, 'collective.contentrules.mail')

PTC_FIXTURE = CCMFixture()
PTC_FUNCTIONAL_TESTING = testing.FunctionalTesting(
    bases=(PTC_FIXTURE, ), name='PloneTestCase:Functional')


class TestCase(PloneTestCase):
    """ Base class used for test cases """

    layer = PTC_FUNCTIONAL_TESTING


collective/contentrules/mail/tests/utils.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
Generic Test case for collective.contentrules.mail doctest
"""
__docformat__ = 'restructuredtext'

from DateTime import DateTime


def addMember(self, username, fullname="", email="", roles=('Member', ), last_login_time=None):
    """Create an new member

    The password is always 'secret'
    """
    self.portal.portal_membership.addMember(username, 'secret', roles, [])
    member = self.portal.portal_membership.getMemberById(username)
    member.setMemberProperties({'fullname': fullname, 'email': email,
                                'last_login_time': DateTime(last_login_time)})


def setUpDefaultMembers(self):
    """Setup default members"""

    addMember(self, 'manager1', 'Manager one', roles=('Manager', 'Member'))
    addMember(self, 'manager2', 'Manager two', roles=('Manager', 'Member'))
    addMember(self, 'reviewer1', 'Reviewer one', roles=('Reviewer', 'Member'))
    addMember(self, 'reviewer2', 'Reviewer two', roles=('Reviewer', 'Member'))
    addMember(self, 'member1', 'Member one', roles=('Member', ))
    addMember(self, 'member2', 'Member two', roles=('Member', ))
    addMember(self, 'authenticated1', 'Authenticated one', roles=('Anonymous', ))
    addMember(self, 'anonymous1', 'Anonymous one', roles=('Anonymous', ))


collective/contentrules/mail/tests/test_maildocs.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
Generic Test case for collective.contentrules.mail doctest
"""
__docformat__ = 'restructuredtext'

import unittest
import doctest
import sys
import os

from Testing.ZopeTestCase import FunctionalDocFileSuite
from Products.Five.testbrowser import Browser

from base import TestCase

current_dir = os.path.dirname(__file__)


def doc_suite(test_dir, setUp=None, tearDown=None, globs=None):
    """Returns a test suite, based on doctests found in /doctest."""
    suite = []
    if globs is None:
        globs = globals()

    # preparing a few elements
    globs['test_dir'] = current_dir
    browser = Browser()
    browser.handleErrors = False
    globs['browser'] = browser

    flags = (doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE |
             doctest.REPORT_ONLY_FIRST_FAILURE |
             doctest.REPORT_UDIFF)

    package_dir = os.path.split(test_dir)[0]
    if package_dir not in sys.path:
        sys.path.append(package_dir)

    docs = []
    for dir_ in ('doctests', 'docs'):
        doctest_dir = os.path.join(package_dir, dir_)

        # filtering files on extension
        docs.extend([os.path.join(doctest_dir, doc) for doc in
                     os.listdir(doctest_dir) if doc.endswith('.txt')])

    for test in docs:
        suite.append(FunctionalDocFileSuite(test, optionflags=flags,
                                            globs=globs, setUp=setUp,
                                            tearDown=tearDown,
                                            test_class=TestCase,
                                            module_relative=False))

    return unittest.TestSuite(suite)


def test_suite():
    """returns the test suite"""
    return doc_suite(current_dir)


collective/contentrules/mail/model.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

from zope.interface import implements
from zope.schema import getFieldsInOrder

from collective.contentrules.mail.interfaces import IMailModel
from collective.contentrules.mail.interfaces import IMailReplacer
from collective.contentrules.mail import MessageFactory as _


class MailModel(object):
    """A mail model described by interface.

    Create instances of this (as shown for DefaultMailModel below) and
    register these as unique named utilities providing IMailModel.
    """

    implements(IMailModel)

    def __init__(self, title, replacer_interface):
        self.title = title
        self.replacer_interface = replacer_interface

    @property
    def fields(self):
        fields = []

        # List of variables provided by replacer interface
        for name, field in getFieldsInOrder(self.replacer_interface):
            fields.append((name, field.title))

        return fields

# The default mail model, which uses the default mail replacer.
DefaultMailModel = MailModel(title=_(u"Standard model for plone content"),
                             replacer_interface=IMailReplacer)


collective/contentrules/mail/config.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

PROJECTNAME = "collective.contentrules.mail"
GLOBALS = globals()


collective/contentrules/mail/__init__.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

# Initialize logger
import logging
from zope.i18nmessageid import MessageFactory as BaseMessageFactory

PROJECTNAME = 'collective.contentrules.mail'
LOG = logging.getLogger(PROJECTNAME)
MessageFactory = BaseMessageFactory(PROJECTNAME)


collective/contentrules/mail/interfaces.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
from zope.interface import Interface
from zope.schema import TextLine, Object, List, Tuple

from collective.contentrules.mail import MessageFactory as _


class IMailModel(Interface):
    """A named utility providing a mail model.

    The mail content rule will allow the user ot pick a mail model based on
    a vocabulary of all named utilities providing this interface. When the
    content rule action is executed, the object being acted upon will be
    adapted to the interface specified under `replacer_interface`.
    Substitutions will then be made based on variables matching field names
    (e.g. "${foobar}" matches the `foobar` field). The substituted values
    are obtained from the adapter.
    """

    title = TextLine(title=_(u"A friendly name for model", ))

    replacer_interface = Object(
        title=_(u"Mail replacer schema"),
        description=_(u"Interface providing word substitution in " \
                        "mail fields: source, recipients, subject, text"),
        schema=Interface)

    fields = List(
        title=_(u"Fields help text"),
        description=_(u"Exposes the variables provided by the replacer"),
        value_type=Tuple(title=_(u"Pair of (key, help text,)"),
                         value_type=TextLine(title=_(u"Name or help text"))))


class IMailReplacer(Interface):
    """Interface providing variables which can be used in mail fields:
    source, recipients, subject, text. This is the default implementation,
    which should work on any CMF content providing the IDublinCore interface.

    It is possible to extend this with other attributes and provide a new
    IMailModel utility with a different interface provided as the
    `replacer_interface`.
    """

    id = TextLine(title=_(u"Id of content", ))

    title = TextLine(title=_(u"Title of content", ))

    description = TextLine(title=_(u"Description of content", ))

    url = TextLine(title=_(u"URL to access content", ))

    relative_url = TextLine(
        title=_(u"Relative URL from portal to access content", ))

    portal_url = TextLine(title=_(u"URL of portal", ))

    owner_id = TextLine(title=_(u"Login of content ower", ))

    owner_fullname = TextLine(title=_(u"Full name of content owner", ))

    owner_emails = TextLine(
        title=_(u"Emails of users having Owner role on content", ))

    reader_emails = TextLine(
            title=_(u"Emails of users having Reader role on content", ))

    contributor_emails = TextLine(
            title=_(u"Emails of users having Contributor role on content", ))

    editor_emails = TextLine(
            title=_(u"Emails of users having Editor role on content", ))

    reviewer_emails = TextLine(
            title=_(u"Emails of users having Reviewer role on content", ))

    default_from_email = TextLine(title=_(u"Email address of default sender", ))

    default_from_name = TextLine(title=_(u"Full name of default sender", ))

    review_state = TextLine(title=_(u"State of content", ))


collective/contentrules/mail/vocabulary.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
from collective.contentrules.mail.interfaces import IMailModel
from collective.contentrules.mail import MessageFactory as _
from zope.component import getUtilitiesFor
from zope.interface import implements
from zope.schema.interfaces import IVocabularyFactory
from zope.schema.vocabulary import SimpleVocabulary, SimpleTerm


class ModelVocabulary(object):
    """Vocabulary factory listing all mail models
    """

    implements(IVocabularyFactory)

    def __call__(self, context):
        terms = [SimpleTerm(name, title=utility.title)
                 for name, utility in getUtilitiesFor(IMailModel)]
        return SimpleVocabulary(terms)


ModelVocabularyFactory = ModelVocabulary()


class MimetypeVocabulary(object):
    """Vocabulary factory listing mail mimetypes
    """

    implements(IVocabularyFactory)

    def __call__(self, context):
        terms = (
            SimpleTerm(
                u"plain",
                title=_(u"text/plain", )),
            SimpleTerm(
                u"html",
                title=_(u"text/html", )),
            )

        return SimpleVocabulary(terms)


MimetypeVocabularyFactory = MimetypeVocabulary()


collective/contentrules/mail/browser/widget.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
from zope.component import getUtility
from zope.app.form.browser.itemswidgets import DropdownWidget
from zope.component.interfaces import ComponentLookupError

from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile

from kss.core import KSSView

from collective.contentrules.mail.interfaces import IMailModel


class ModelKSSView(KSSView):
    help_template = ViewPageTemplateFile("templates/model.pt")

    def updateModelDescription(self, model_id=''):
        # Get model and display help
        try:
            model = getUtility(IMailModel, model_id)
        except ComponentLookupError:
            model = None

        help = ""

        if model is not None:
            help = self.help_template(fields=model.fields)

        core = self.getCommandSet('core')
        core.replaceInnerHTML('#mail-model .modelDescription', help)
        return self.render()


class ModelWidget(DropdownWidget):

    help_template = ViewPageTemplateFile("templates/model.pt")

    def __init__(self, field, request):
        """Initialize the widget."""
        super(ModelWidget, self).__init__(field,
            field.vocabulary, request)

    def __call__(self):
        """See IBrowserWidget."""
        value = self._getFormValue()
        contents = []
        contents.append(self._div('value', self.renderValue(value)))
        contents.append(self._emptyMarker())

        # Get model and display model description
        # If no model selected, get the first value of vocabulary
        help = ""

        if not value and len(self.vocabulary):
            value = self.vocabulary._terms[0].value

        # Get model
        if value:
            try:
                model = getUtility(IMailModel, value)
            except ComponentLookupError:
                model = None

            if model is not None:
                help = self.help_template(fields=model.fields)

        contents.append(self._div("modelDescription", help))
        return self._div(self.cssClass, "\n".join(contents), id="mail-model")


collective/contentrules/mail/actions/mail.py
# -*- coding: utf-8 -*-
# Copyright (c) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
from Acquisition import aq_inner
from OFS.SimpleItem import SimpleItem
from zope.component import adapts
from zope.component import getUtility
from zope.component.interfaces import ComponentLookupError
from zope.interface import Interface
from zope.interface import implements
from zope.formlib import form
from zope.schema import TextLine
from zope.schema import Text
from zope.schema import Choice
from zope.schema import getFieldNamesInOrder

from Products.CMFCore.utils import getToolByName

from plone.app.contentrules.browser.formhelper import AddForm
from plone.app.contentrules.browser.formhelper import EditForm
from plone.contentrules.rule.interfaces import IRuleElementData
from plone.contentrules.rule.interfaces import IExecutable

from collective.contentrules.mail import MessageFactory as _
from collective.contentrules.mail import LOG
from collective.contentrules.mail.browser.widget import ModelWidget
from collective.contentrules.mail.interfaces import IMailModel

from smtplib import SMTPException
from Products.MailHost.MailHost import MailHostError


class IMailAction(Interface):
    """Definition of the configuration available for a mail action
    """

    model = Choice(
        title=_(u"Mail model"),
        required=True,
        vocabulary="collective.contentrules.mail.vocabulary.model", )

    mimetype = Choice(
        title=_(u"Mail mimetype"),
        required=True,
        vocabulary="collective.contentrules.mail.vocabulary.mimetype", )

    subject = TextLine(
        title=_(u"Subject"),
        description=_(u"Subject of the message"),
        required=True)

    source = TextLine(
        title=_(u"Email source"),
        description=_("The email address that sends the email. If no email is" \
                      " provided here, it will use the portal from address."),
        required=False)

    recipients = TextLine(
        title=_(u"Email recipients"),
        description=_("The email where you want to send this message. To send" \
            " it to different email addresses, just separate them with commas."),
        required=True)

    cc = TextLine(
        title=_(u"CC recipients"),
        description=_("The email to receive a copy of this message. To send" \
            " it to different email addresses, just separate them with commas."),
        required=False)

    bcc = TextLine(
        title=_(u"BCC recipients"),
        description=_("The email to receive a blind copy of this message. To" \
     " send it to different email addresses, just separate them with commas."),
        required=False)

    message = Text(
        title=_(u"Message"),
        description=_(u"Type in here the message that you want to mail."),
        required=True)


class MailAction(SimpleItem):
    """
    The implementation of the action defined before
    """
    implements(IMailAction, IRuleElementData)

    model = ''
    mimetype = ''
    subject = ''
    source = ''
    recipients = ''
    cc = ''
    bcc = ''
    message = ''

    element = 'collective.contentrules.mail.actions.Mail'

    @property
    def summary(self):
        return _(u"Email report to ${recipients}",
                 mapping=dict(recipients=self.recipients))


class MailActionExecutor(object):
    """The executor for this action.
    """
    implements(IExecutable)
    adapts(Interface, IMailAction, Interface)

    def __init__(self, context, element, event):
        self.context = context
        self.element = element
        self.event = event

    def __call__(self):
        # Get event objet
        obj = self.event.object

        utool = getToolByName(aq_inner(self.context), "portal_url")
        portal = utool.getPortalObject()

        site_charset = 'utf-8'
        email_charset = portal.getProperty('email_charset')

        # Get replacer interface from model
        model = getUtility(IMailModel, self.element.model)
        replacer_interface = model.replacer_interface

        # Extract all variables from this replacer
        replacer = replacer_interface(obj, None)
        if replacer is None:
            LOG.debug(u"Could not send email. The replacer is not applicable for this type of object.")
            return False

        words = {}
        for word_id in getFieldNamesInOrder(replacer_interface):
            replacement_value = getattr(replacer, word_id)
            if isinstance(replacement_value, str):
                replacement_value = replacement_value.decode(site_charset)
            words[word_id] = replacement_value

        # Apply word substitution on every mail fields
        def substitute(text):
            if not text:
                return ""

            for word_id, replacement_value in words.items():
                text = text.replace(u"${%s}" % word_id, replacement_value)

            return text.encode(site_charset)

        source = self.element.source
        if source:
            source = substitute(self.element.source)
        recipients = substitute(self.element.recipients)
        cc = substitute(self.element.cc)
        bcc = substitute(self.element.bcc)
        subject = substitute(self.element.subject)
        message = substitute(self.element.message)

        def processRecipients(recipients=''):
            address_list = []
            for email in recipients.split(','):
                email = email.strip()
                if not email:
                    # Remove empty address
                    continue
                address_list.append(email)
            return address_list

        # Process recipients
        recipient_list = processRecipients(recipients)
        cc_list = [e for e in processRecipients(cc) if e not in recipients]
        bcc_list = [e for e in processRecipients(bcc) if e not in recipients]

        if not recipient_list:
            # Because there are no recipients, do not send email
            LOG.info(u"""Do not send email "%s": no recipients defined.""" %
                     model.title)
            return False

        recipients = ",".join(recipient_list)
        cc = ",".join(cc_list)
        bcc = ",".join(bcc_list)

        # Process source

        if not source:
            # no source provided, looking for the site wide from email
            # address
            from_address = portal.getProperty('email_from_address')

            if not from_address:
                raise ValueError, "You must provide a source address for this\
action or enter an email in the portal properties"

            from_name = portal.getProperty('email_from_name')
            source = "%s <%s>" % (from_name, from_address)

        # Encode text using mail charset

        if email_charset != site_charset:
            source = source.decode(site_charset).encode(email_charset)
            recipients = recipients.decode(site_charset).encode(email_charset)
            cc = cc.decode(site_charset).encode(email_charset)
            bcc = bcc.decode(site_charset).encode(email_charset)
            subject = subject.decode(site_charset).encode(email_charset)
            message = message.decode(site_charset).encode(email_charset)

        # Send email
        mailhost = getToolByName(aq_inner(self.context), "MailHost")

        if not mailhost:
            raise ComponentLookupError, \
                "You must have a Mailhost utility to execute this action"

        mimetype = self.element.mimetype

        try:
            mailhost.secureSend(message, recipients, source,
                                subject=subject, subtype=mimetype,
                                mcc=cc, mbcc=bcc,
                                charset=email_charset, debug=False)
        except (MailHostError, SMTPException, ), e:
            LOG.exception(u"Failed to send mail")
            return False

        return True


class MailAddForm(AddForm):
    """
    An add form for the mail action
    """
    form_fields = form.FormFields(IMailAction)
    form_fields['model'].custom_widget = ModelWidget
    label = _(u"Add Mail Action")
    description = _(u"A mail action can mail different recipient.")
    form_name = _(u"Configure element")

    def create(self, data):
        a = MailAction()
        form.applyChanges(a, self.form_fields, data)
        return a


class MailEditForm(EditForm):
    """
    An edit form for the mail action
    """
    form_fields = form.FormFields(IMailAction)
    form_fields['model'].custom_widget = ModelWidget
    label = _(u"Edit Mail Action")
    description = _(u"A mail action can mail different recipient.")
    form_name = _(u"Configure element")
